zkevm-circuits/src/blake2b/src/main.rs

```diff
@@ -0,0 +1,442 @@
+    for r in 0..=max {
+        region.assign_fixed(|| format!("[0..{}] lookup table, row {}", max, r), table, r as usize, || Value::known(F::from(r as u64)))?;
+    }
+    return Ok(());
+}
+
+fn constrain_to_set<F:FieldExt>(meta: &mut ConstraintSystem<F>, table: Column<Fixed>, selector: Column<Fixed>, cell: ACell<F>) {
+    meta.lookup_any("Set constraint", |meta| {
+        let selector = meta.query_fixed(selector, Rotation::cur());
+        return vec![(selector.clone() * cell.express(meta), meta.query_fixed(table, Rotation::cur()))];
+    });
+}
+
+fn constrain_to_bit<F:FieldExt>(meta: &mut ConstraintSystem<F>, selector: Column<Fixed>, cell: ACell<F>) {
+    meta.create_gate("Bit constraint", |meta| {
+        let selector = meta.query_fixed(selector, Rotation::cur());
+        let bit = cell.express(meta);
+        return vec![selector * bit.clone() * (gf(1) - bit)];
+    });
+}
+
+fn set_byte_xor_lookup<F:FieldExt>(region: &mut Region<F>, table: &[Column<Fixed>; 3]) -> Result<(), Error> {
+    for r in 0..65536 {
+        let (first, second) = (r as u64 & 0xFF, r as u64 >> 8);
+        region.assign_fixed(|| format!("Xor lookup table, row {}, the first", r), table[0], r, || Value::known(F::from(first)))?;
+        region.assign_fixed(|| format!("Xor lookup table, row {}, the second", r), table[1], r, || Value::known(F::from(second)))?;
+        region.assign_fixed(|| format!("Xor lookup table, row {}, the result", r), table[2], r, || Value::known(F::from(first ^ second)))?;
+    }
+    return Ok(());
+}
+
+fn constrain_to_byte_xor<F:FieldExt>(meta: &mut ConstraintSystem<F>, table: &[Column<Fixed>; 3], 
+    selector: Column<Fixed>, first: ACell<F>, second: ACell<F>, result: ACell<F>) {
+        meta.lookup_any("Byte xor constraint", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            return vec![
+                (selector.clone() * first.express(meta), meta.query_fixed(table[0], Rotation::cur())),
+                (selector.clone() * second.express(meta), meta.query_fixed(table[1], Rotation::cur())),
+                (selector.clone() * result.express(meta), meta.query_fixed(table[2], Rotation::cur()))];
+        });
+}
+
+fn set_result_and_activate<F:FieldExt, const N:usize >(region: &mut Region<F>, name: &str, 
+    row: usize, selector: Column<Fixed>, cells: &[ACell<F>; N], result: u128) -> Result<(), Error> {
+        set_byte_witnesses(region, &format!("{}, result", name), row, cells, result.to_le_bytes()[0..N].try_into().unwrap())?;     
+        region.assign_fixed(|| format!("{}, selector", name), selector, row, || Value::known(F::one()))?;
+        return Ok(());
+}
+
+#[derive(Debug, Copy, Clone)]
+struct ACell<F:FieldExt> {
+    pub column: Column<Advice>,
+    pub offset: u16,
+    _marker: PhantomData<F>
+}
+
+impl<F:FieldExt> ACell<F> {
+    fn express(&self, meta: &mut VirtualCells<F>) -> Expression<F> {
```

Comment:
```suggestion
    fn expr(&self, meta: &mut VirtualCells<F>) -> Expression<F> {
```
Just because this is the way it's called in most other code in the project.